lib/vector-core/src/config/mod.rs

```diff
@@ -171,3 +176,91 @@ impl From<bool> for AcknowledgementsConfig {
         Some(enabled).into()
     }
 }
+
+#[derive(Debug, Copy, Clone, PartialEq, Serialize, Deserialize, PartialOrd, Ord, Eq)]
+pub enum LogNamespace {
+    /// Vector native namespacing
+    ///
+    /// Deserialized data is placed in the root of the event.
+    /// Extra data is placed in "event metadata"
+    Vector,
+
+    /// This is the legacy namespacing.
+    ///
+    /// All data is set in the root of the event. Since this can lead
+    /// to collisions, deserialized data has priority over metadata
+    Legacy,
+}
+
+impl From<bool> for LogNamespace {
+    fn from(x: bool) -> Self {
+        if x {
+            LogNamespace::Vector
+        } else {
+            LogNamespace::Legacy
+        }
+    }
+}
+
+impl Default for LogNamespace {
+    fn default() -> Self {
+        Self::Legacy
+    }
+}
+
+impl LogNamespace {
+    /// Vector: This is added to "event metadata", nested under the source name
+    ///
+    /// Legacy: This is stored on the event root, only if a field with that name doesn't already exist
+    pub fn insert_source_metadata<'a>(
+        &self,
+        source_name: &'a str,
+        log: &mut LogEvent,
+        key: impl Path<'a>,
+        value: impl Into<Value>,
+    ) {
+        match self {
+            LogNamespace::Vector => {
+                log.metadata_mut()
+                    .value_mut()
+                    .insert(path!(source_name).concat(key), value);
+            }
+            LogNamespace::Legacy => {
+                log.try_insert(key, value);
+            }
+        }
+    }
+
+    /// Vector: This is added to the "event metadata", nested under the name "vector". This data
+    /// will be marked as read-only in VRL
+    ///
+    /// Legacy: This is stored on the event root, only if a field with that name doesn't already exist
+    pub fn insert_vector_metadata<'a>(
+        &self,
+        log: &mut LogEvent,
+        legacy_key: impl Path<'a>,
+        metadata_key: impl Path<'a>,
+        value: impl Into<Value>,
+    ) {
+        match self {
+            LogNamespace::Vector => {
+                log.metadata_mut()
+                    .value_mut()
+                    .insert(path!("vector").concat(metadata_key), value);
+            }
+            LogNamespace::Legacy => log.try_insert(legacy_key, value),
+        }
+    }
+
+    pub fn new_log_from_data(&self, value: impl Into<Value>) -> LogEvent {
+        match self {
+            LogNamespace::Vector | LogNamespace::Legacy => LogEvent::from(value.into()),
+        }
+    }
+
+    // combine a global (self) and local value to get the actual namespace
+    #[must_use]
+    pub fn merge(&self, override_value: Option<impl Into<LogNamespace>>) -> LogNamespace {
+        override_value.map_or(*self, Into::into)
+    }
```

Comment:
This is just a convenience function so you can merge the global `log_namespace` value with the source value. It will need to be used in every source, so I wanted it to be clear what's going on.

> I think the caller should handle this no-op instead

Right now, it's usage is pretty self-documenting. If the caller had to handle this each time it would be a bit more confusing, especially considering the source config is a `bool`, and it needs to be converted to a `LogNamespace` before it's actually merged. 
```rust
/// This is in the `outputs` function of a source component

let log_namespace = global_log_namespace.merge(self.log_namespace);
```